bench/lmdb_bench.ml

```diff
@@ -0,0 +1,237 @@
+
+  let fail_on_error f =
+    match f () with Ok _ -> () | Error err -> failwith (string_of_error err)
+
+  let flags = [ Lmdb.NoRdAhead; Lmdb.NoSync; Lmdb.NoMetaSync; Lmdb.NoTLS ]
+
+  let mapsize = 409_600_000_000L
+
+  let get_wtxn dir flags =
+    cleanup ();
+    opendir dir ~mapsize ~flags 0o644 >>= fun env ->
+    create_rw_txn env >>= fun txn ->
+    opendb txn >>= fun ddb -> Ok ((txn, ddb), env)
+
+  let write (txn, ddb) () =
+    Array.iter
+      (fun (k, v) -> fail_on_error (fun () -> Lmdb.put_string txn ddb k v))
+      random
+
+  let read (txn, ddb) () =
+    Array.iter
+      (fun (k, _) ->
+        ignore (Bigstring.to_string (R.get_ok (Lmdb.get txn ddb k))))
+      random
+
+  let write_random () =
+    get_wtxn root flags >>| fun (rw, env) ->
+    print_results (write rw) nb_entries;
+    print_stats rw;
+    (rw, env)
+
+  let write_seq () =
+    Array.sort (fun a b -> String.compare (fst a) (fst b)) random;
+    get_wtxn root flags >>| fun (rw, env) ->
+    print_results (write rw) nb_entries;
+    closedir env
+
+  let write_sync () =
+    get_wtxn root [ Lmdb.NoRdAhead ] >>| fun (rw, env) ->
+    let write (txn, ddb) env ls () =
+      Array.iter
+        (fun (k, v) ->
+          fail_on_error (fun () ->
+              Lmdb.put_string txn ddb k v >>= fun () -> sync env))
+        ls
+    in
+    print_results (write rw env random) nb_entries;
+    closedir env
+
+  let overwrite rw = print_results (write rw) nb_entries
+
+  let read_random r = print_results (read r) nb_entries
+
+  (*use a new db, created without the flag Lmdb.NoRdAhead*)
+  let read_seq () =
+    let rw, env =
+      R.get_ok
+        ( get_wtxn root [ Lmdb.NoSync; Lmdb.NoMetaSync ] >>| fun (rw, env) ->
+          let () = write rw () in
+          (rw, env) )
+    in
+    let read (txn, ddb) () =
+      opencursor txn ddb >>= fun cursor ->
+      cursor_first cursor >>= fun () ->
+      cursor_iter
+        ~f:(fun (k, v) ->
+          ignore (Bigstring.to_string k);
+          ignore (Bigstring.to_string v);
+          Ok ())
+        cursor
+      >>| fun () -> cursor_close cursor
+    in
+    let aux_read r () = fail_on_error (read r) in
+    print_results (aux_read rw) nb_entries;
+    closedir env
+
+  let close env = closedir env
+end
+
+let init () =
+  Common.report ();
+  Lmdb.cleanup ();
+  Log.app (fun l -> l "Keys: %d bytes each." key_size);
+  Log.app (fun l -> l "Values: %d bytes each." value_size);
+  Log.app (fun l -> l "Entries: %d." nb_entries);
+  Log.app (fun l -> l "Log size: %d." log_size);
+  populate ()
+
+let run input =
+  init ();
+  Log.app (fun l -> l "\n Fill in random order");
+  let lmdb, env = R.get_ok (Lmdb.write_random ()) in
+  let match_input ~bench ~triggers ~message =
+    List.iter
+      (fun trigger ->
+        if input = trigger then
+          let () = Log.app (fun l -> l "\n %s" message) in
+          bench ()
+        else ())
+      triggers
+  in
+  let () =
+    match_input
```

Comment:
Nice :slightly_smiling_face: Now you can keep `(unit -> unit, trigger list, string)` tuples in a list and do something like:
```ocaml
List.iter (fun (bench, triggers, message) -> match_input ~bench ~triggers ~message)
```
Keeping specifications in data rather than code etc. etc.